### PR TITLE
Delete route

### PR DESCRIPTION
**Pull request by noctera**
| Status  | Type  | Env Vars Change | 
| :---: | :---: | :---: |
| Ready | Feature | Yes |

## Description

Routes to delete user, language package, group, vocabulary cards 

New DELETE routes:

- `/api/vocabulary/:vocabularyId`
- `/api/group/:groupId`
- `/api/languagePackage/:languagePackageId`
- `/api/user`

## Motivation and Context

This PR is needed to delete user, language package, group, vocabulary cards 

## How Has This Been Tested?

locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking change which cleans up / improves existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the **CONTRIBUTING** document.
- [x] I have added unit tests to code changes that make sense.
- [x] I have considered the accessibility of my changes (i.e. did I add proper content descriptions to images, or run my changes with talkback enabled?)
- [x] I have documented my code if needed

### CODE REVIEW

File: app/Controllers/AuthController.js [hunk 1]
```diff
@@ -31,8 +37,18 @@ async function profile(req, res) {
   res.send(deleteKeysFromObject(['roleId', 'password', 'createdAt', 'updatedAt'], req.user.toJSON()));
 }
 
+async function deleteUser(req, res) {
+  // get userId from request
+  const userId = req.user.id;
+
+  destroyUser(userId);
```
Comment (luwol03) [inline app/Controllers/AuthController.js]: add `await`

File: app/Controllers/AuthController.js [hunk 2]
```diff
@@ -31,8 +37,18 @@ async function profile(req, res) {
...
+
+  res.sendStatus(200);
```
Comment (luwol03) [inline app/Controllers/AuthController.js]: send 204

File: app/Controllers/GroupController.js [hunk 3]
```diff
@@ -1,8 +1,8 @@
-const { createGroup, getGroups } = require('../Services/GroupServiceProvider.js');
+const { createGroup, getGroups, destroyGroup } = require('../Services/GroupServiceProvider.js');
 
 async function addGroup(req, res) {
   // get userId from request
-  const { id } = await req.user;
+  const { id } = req.user;
```
Comment (luwol03) [inline app/Controllers/GroupController.js]: could we use 
```js
const userId = req.user.id;
```
for consistency by getting the userId?

File: app/Controllers/GroupController.js [hunk 4]
```diff
@@ -26,7 +26,18 @@ async function sendGroups(req, res) {
   res.send(groups);
 }
 
+async function deleteGroup(req, res) {
+  // get userId from request
+  const userId = req.user.id;
+  const { groupId } = req.params;
+
+  destroyGroup(userId, groupId);
```
Comment (luwol03) [inline app/Controllers/GroupController.js]: also missing `await`

File: app/Controllers/LanguagePackageController.js [hunk 5]
```diff
@@ -22,26 +27,55 @@ async function addLanguagePackage(req, res) {
 
 async function sendLanguagePackages(req, res) {
   // get userId from request
-  const { id } = await req.user;
+  const { id } = req.user;
+  const { groups } = req.query || false;
```
Comment (luwol03) [inline app/Controllers/LanguagePackageController.js]: this makes no sense to destructure a boolean value

File: app/Controllers/LanguagePackageController.js [hunk 6]
```diff
@@ -22,26 +27,55 @@ async function addLanguagePackage(req, res) {
...
 
   // get language Package
   const languagePackages = await getLanguagePackages(id, res);
```
Comment (luwol03) [inline app/Controllers/LanguagePackageController.js]: could you add a parameter groups to this function to decide if we wan't to include the groups insted of this messy if solution.

```diff
diff --git a/app/Controllers/LanguagePackageController.js b/app/Controllers/LanguagePackageController.js
index 3a54942..6189a50 100644
--- a/app/Controllers/LanguagePackageController.js
+++ b/app/Controllers/LanguagePackageController.js
@@ -5,7 +5,6 @@ const {
 } = require('../Services/LanguagePackageServiceProvider.js');
 const { createDrawers } = require('../Services/DrawerServiceProvider.js');
 const { drawers } = require('../utils/constants.js');
-const { getGroups } = require('../Services/GroupServiceProvider.js');
 const {
   getNumberOfUnresolvedVocabulary,
   getNumberOfUnactivatedVocabulary,
@@ -31,35 +30,18 @@ async function sendLanguagePackages(req, res) {
   const { groups } = req.query || false;

   // get language Package
-  const languagePackages = await getLanguagePackages(id, res);
+  const languagePackages = await getLanguagePackages(id, res, groups);

-  let formatted;
-  // if groups is true, return groups to every language package
-  if (groups) {
-    formatted = await Promise.all(
-      languagePackages.map(async (languagePackage) => ({
-        unresolvedVocabularies: await getNumberOfUnresolvedVocabulary(languagePackage.id, id),
+  const formatted = await Promise.all(
+    languagePackages.map(async (languagePackage) => ({
+      unresolvedVocabularies: await getNumberOfUnresolvedVocabulary(languagePackage.id, id),

-        // add number of unactivated vocabularies
-        unactivatedVocabularies: await getNumberOfUnactivatedVocabulary(languagePackage.id, id),
+      // add number of unactivated vocabularies
+      unactivatedVocabularies: await getNumberOfUnactivatedVocabulary(languagePackage.id, id),

-        groups: await getGroups(id, languagePackage.id, res),
-
-        ...languagePackage.toJSON(),
-      }))
-    );
-  } else {
-    formatted = await Promise.all(
-      languagePackages.map(async (languagePackage) => ({
-        unresolvedVocabularies: await getNumberOfUnresolvedVocabulary(languagePackage.id, id),
-
-        // add number of unactivated vocabularies
-        unactivatedVocabularies: await getNumberOfUnactivatedVocabulary(languagePackage.id, id),
-
-        ...languagePackage.toJSON(),
-      }))
-    );
-  }
+      ...languagePackage.toJSON(),
+    }))
+  );

   res.send(formatted);
 }
diff --git a/app/Services/LanguagePackageServiceProvider.js b/app/Services/LanguagePackageServiceProvider.js
index 6ce7316..d529e3e 100644
--- a/app/Services/LanguagePackageServiceProvider.js
+++ b/app/Services/LanguagePackageServiceProvider.js
@@ -1,4 +1,4 @@
-const { LanguagePackage } = require('../../database');
+const { LanguagePackage, Group } = require('../../database');

 // create language package
 async function createLanguagePackage(
@@ -18,9 +18,10 @@ async function createLanguagePackage(
 }

 // get language package
-async function getLanguagePackages(userId, res) {
+async function getLanguagePackages(userId, res, groups) {
   // Get user with email from database
   const languagePackages = await LanguagePackage.findAll({
+    include: groups ? [{ model: Group, attributes: ['id', 'name', 'active'] }] : [],
     attributes: ['id', 'name', 'foreignWordLanguage', 'translatedWordLanguage', 'vocabsPerDay', 'rightWords'],
     where: {
       userId: userId,
```

File: app/Controllers/LanguagePackageController.js [hunk 7]
```diff
@@ -22,26 +27,55 @@ async function addLanguagePackage(req, res) {
...
 
   res.send(formatted);
 }
 
+async function deleteLanguagePackage(req, res) {
+  // get userId from request
+  const userId = req.user.id;
+  const { languagePackageId } = req.params;
+
+  destroyLanguagePackage(userId, languagePackageId);
```
Comment (luwol03) [inline app/Controllers/LanguagePackageController.js]: also missing `await`

File: app/Controllers/LanguagePackageController.js [hunk 8]
```diff
@@ -22,26 +27,55 @@ async function addLanguagePackage(req, res) {
...
+
+  res.sendStatus(200);
```
Comment (luwol03) [inline app/Controllers/LanguagePackageController.js]: 204

File: app/Controllers/VocabularyController.js [hunk 9]
```diff
@@ -18,6 +23,28 @@ async function addVocabularyCard(req, res) {
   res.sendStatus(204);
 }
 
+async function deleteVocabularyCard(req, res) {
+  // get userId from request
+  const userId = req.user.id;
+  const { vocabularyId } = req.params;
+
+  destroyVocabularyCard(userId, vocabularyId);
```
Comment (luwol03) [inline app/Controllers/VocabularyController.js]: add `await`

File: app/Controllers/AuthController.js [hunk 10]
```diff
@@ -31,8 +37,18 @@ async function profile(req, res) {
   res.send(deleteKeysFromObject(['roleId', 'password', 'createdAt', 'updatedAt'], req.user.toJSON()));
 }
 
+async function deleteUser(req, res) {
+  // get userId from request
+  const userId = req.user.id;
```
Comment (luwol03) [inline app/Controllers/AuthController.js:42]: shouldn't we add a security check so that a user have to send also his password?

**Review by luwol03 — CHANGES_REQUESTED** [review]

File: app/Controllers/AuthController.js (hunk 10 above)
Comment (noctera) [reply app/Controllers/AuthController.js:42]: will be added in another PR

**Review by luwol03 — APPROVED** [review]

**Pull request merged**